### PR TITLE
New chain params system

### PR DESCRIPTION
**Pull request by edfelten**
This creates a new system for ArbOS/chain parameters.  A global map stores paramId -> value mappings, where paramId is the keccak256 hash of the parameter's name.  All parameters are stored as uint values.

Parameters are listed, and given default values, in the `constants.json` file.  Parameters are initialized to their default values when ArbOS boots.

The format of the chain initialization message changes.  In the new format, the type-specific data consists of a sequence of zero or more 64-byte chunks, where each chunk is a pair (paramId, value).

New ArbOwner methods support getting and setting parameters. Old parameter-specific getters and setters are removed.

This will fail CI.  In particular, the `evm::preinstalled_contracts::_test_upgrade_arbos_to_different_version` test will fail, because the test harness sends a new-style init message when starting the arbos_before.mexe, but that old version of ArbOS expects an old-style init message.  (This issue will not come up when upgrading on the real chain, because it is already booted.). We need to rely on integration testing to make sure that the upgrade succeeds.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -305,6 +313,19 @@ fn main() -> Result<(), CompileError> {
...
         }
+        Args::MakeParametersList(clist) => {
+            let constants_map =
+                make_parameters_list(clist.consts_file.as_ref().map(|s| Path::new(s))).unwrap();
+            match serde_json::to_string(&constants_map) {
+                Ok(s) => {
+                    println!("{}", s);
+                }
+                Err(e) => {
+                    panic!("{}", e);
```
Comment (PlasmaPower) [inline src/main.rs:324]: If you're going to panic with the error, you might as well call unwrap on the result.

**Review by PlasmaPower — CHANGES_REQUESTED** [review]

File: arb_os/arbowner.mini [hunk 2]
```diff
@@ -535,6 +499,63 @@ impure func arbowner_setL1GasPriceEstimate(_topFrame: EvmCallFrame, calldata: By
     evmOp_return(0, 0);
 }
 
+impure func arbowner_setFairGasPriceSender(_topFrame: EvmCallFrame, calldata: ByteArray) { // (address)
```
Comment (PlasmaPower) [inline arb_os/arbowner.mini:502]: This was just moved around in this PR, but this comment is wrong. The params are `(address, bool)`.

File: arb_os/arbowner.mini [hunk 3]
```diff
@@ -396,6 +335,31 @@ impure func arbowner_deployContract(topFrame: EvmCallFrame, calldata: ByteArray)
...
+    if (bytearray_size(calldata) != 4) {
+        evmOp_revert_knownPc(170, 0, 0);
+    }
+    let serialized = chainParams_serializeAllParams();
+    let serSize = bytearray_size(serialized);
+    let mem = bytearray_copy(
+        serialized,
+        0,
+        bytearray_set256(
+            bytearray_set256(bytearray_new(0), 0, 64),
```
[the remark this answers is not in the document]
Comment (edfelten) [reply arb_os/arbowner.mini:349]: The ultimate fix for this is the antigen tool, which is in our pipeline.  So I prefer to leave this as is for now.

File: arb_os/chainParameters.mini [hunk 4]
```diff
@@ -1,208 +1,144 @@
...
 }
 
-public impure func chainParams_chainId() -> uint {
-    return xif let Some(params) = globalChainParams {
-        params.chainId
-    } else {
-        const::DefaultChainID
-    };
+public impure func chainParams_set(which: uint, value: uint) {
+    globalChainParameters = globalChainParameters with { [which] = value };
```
[the remark this answers is not in the document]
Comment (edfelten) [reply arb_os/chainParameters.mini:36]: Changed this to validate that the param already exists.  Added a new chainParams_create which sets the value regardless of whether it existed before.

File: arb_os/chainParameters.mini [hunk 5]
```diff
@@ -1,208 +1,144 @@
...
+            0
+        };
+        let value = xif let Some(res) = bytestream_get256(stream) {
+            let (ustr, val) = res;
+            stream = ustr;
+            val
+        } else {
+            notifyParamsChanged();
+            return;
+            0
```
[the remark this answers is not in the document]
Comment (edfelten) [reply arb_os/chainParameters.mini:96]: Agree. Changed it.

File: arb_os/chainParameters.mini (hunk 4 above)
[the remark this answers is not in the document]
Comment (PlasmaPower) [reply arb_os/chainParameters.mini:36]: This fix and the one for chainParams_gotParamsMessage don't seem to have been pushed up yet

**Comment by RachelFranks** [comment]
This has been merged

**Pull request closed without merging**